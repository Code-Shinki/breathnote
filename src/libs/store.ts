import axios from 'axios'
import { AllKeywords } from 'types/keyword'
import { AllPosts } from 'types/post'
import { API_ENDPOINT, API_KEY } from 'utils/env'

const KEY = { headers: { 'X-API-KEY': API_KEY } }

const fetchAllPosts = async () => {
  console.log('ALL_POSTS fire')

  // 一度に受け取る記事数
  const limit = 10

  // 総記事数等を含む最初の10記事を取得
  const posts: AllPosts = await axios
    .get(`${API_ENDPOINT}/posts?limit=${limit}`, KEY)
    .then((res) => res.data)
    .catch((err) => {
      throw new Error(`FETCH FAILED (fetchAllPosts() - First reqest) : ${err}`)
    })

  // APIを叩く回数
  const requestsCount = Math.floor(posts.totalCount / limit)

  // すべての記事を取得してposts.contentsに追加
  for (let i = 1; i <= requestsCount; i++) {
    const fetchData = await axios
      .get(`${API_ENDPOINT}/posts?offset=${limit * i}&limit=${limit}`, KEY)
      .then((res) => res.data)
      .catch((err) => {
        throw new Error(`FETCH FAILED (fetchAllPosts() - Roop ${i} reqest) : ${err}`)
      })
    posts.contents = posts.contents.concat(fetchData.contents)
  }

  // 記事を降順 (公開日が新しい順) に並び替え
  if (posts.contents.length > 1) {
    posts.contents.sort((a, b) => {
      if (a.publishedAt < b.publishedAt) {
        return 1
      } else {
        return -1
      }
    })
  }

  return posts
}

const fetchAllKeywords = async () => {
  console.log('ALL_KEYWORDS fire')

  // 一度に受け取るキーワード数
  const limit = 10

  // 総キーワード数等を含む最初の10データを取得
  const keywords: AllKeywords = await axios
    .get(`${API_ENDPOINT}/keywords?limit=${limit}`, KEY)
    .then((res) => res.data)
    .catch((err) => {
      throw new Error(`FETCH FAILED (fetchAllKeywords() - First reqest) : ${err}`)
    })

  // APIを叩く回数
  const requestsCount = Math.floor(keywords.totalCount / limit)

  // すべてのキーワードを取得してposts.contentsに追加
  for (let i = 1; i <= requestsCount; i++) {
    const fetchData = await axios
      .get(`${API_ENDPOINT}/keywords?offset=${limit * i}&limit=${limit}`, KEY)
      .then((res) => res.data)
      .catch((err) => {
        throw new Error(`FETCH FAILED (fetchAllKeywords() - Roop ${i} reqest) : ${err}`)
      })
    keywords.contents = keywords.contents.concat(fetchData.contents)
  }

  // キーワードを昇順 (アルファベット順) に並び替え
  if (keywords.contents.length > 1) {
    keywords.contents.sort((a, b) => {
      if (a.name > b.name) {
        return 1
      } else {
        return -1
      }
    })
  }

  return keywords
}

export const ALL_POSTS = fetchAllPosts()
export const ALL_KEYWORDS = fetchAllKeywords()
