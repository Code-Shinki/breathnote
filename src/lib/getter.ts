/* ＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/

  LIB -> GETTER

  - STORE から非同期処理で取得したデータを返却

＿/ ＿/ ＿/ ＿/ ＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/ */

import * as Store from 'lib/store'
import { CategoryType, ConfigType, PostType, TagType } from 'lib/types'

/* -------------------------------------------
  getConfig : CONFIG から必要なデータを抜き出して返却
------------------------------------------- */
export const getConfig = async (): Promise<ConfigType> => {
  const allConfig = await Store.CONFIG
  const config = {
    siteDomain: allConfig.siteDomain,
    siteTitle: allConfig.siteTitle,
    siteSubTitle: allConfig.siteSubTitle,
    siteDescription: allConfig.siteDescription,
    siteKeywords: allConfig.siteKeywords,
  }

  return config
}

/* -------------------------------------------
  getSortedPosts : ALL_POSTS を日付が新しい順に並べ替えて返却
  getAllPostPaths : ALL_POSTS から slug を抜き出して返却
  getPost : 受け取った slug に一致する記事を ALL_POSTS から返却
------------------------------------------- */
export const getSortedPosts = async (): Promise<PostType[]> => {
  const allPosts = await Store.ALL_POSTS

  return allPosts.sort((a, b) => {
    if (a.publishedAt < b.publishedAt) {
      return 1
    } else {
      return -1
    }
  })
}

export const getAllPostPaths = async (): Promise<{ params: { slug: string } }[]> => {
  const allPosts = await Store.ALL_POSTS

  return allPosts.map((post) => {
    return {
      params: {
        slug: post.slug,
      },
    }
  })
}

export const getPost = async (slug: string): Promise<PostType> => {
  const allPosts = await Store.ALL_POSTS
  const post = allPosts.find((post) => post.slug === slug)

  if (post) {
    return post
  } else {
    throw 'Post not found.'
  }
}

/* -------------------------------------------
  getAllCategories : ALL_CATEGORIES を返却
  getAllCategoryPaths : ALL_CATEGORIES から slug を抜き出して返却
  getCategoryPosts : 受け取った slug に一致する記事を ALL_POSTS から返却
  getCategory : 受け取った slug に一致するカテゴリを ALL_CATEGORIES から返却
------------------------------------------- */
export const getAllCategories = async (): Promise<CategoryType[]> => {
  const allCategories = await Store.ALL_CATEGORIES

  return allCategories
}

export const getAllCategoryPaths = async (): Promise<{ params: { slug: string } }[]> => {
  const allCategories = await Store.ALL_CATEGORIES

  return allCategories.map((category) => {
    return {
      params: {
        slug: category.slug,
      },
    }
  })
}

export const getCategoryPosts = async (slug: string): Promise<PostType[]> => {
  const allPosts = await Store.ALL_POSTS
  const categoryPosts = allPosts.filter((post) => {
    const isFound = post.categories.some((category) => {
      return category.slug === slug
    })
    return isFound
  })

  return categoryPosts
}

export const getCategory = async (slug: string): Promise<CategoryType> => {
  const allCategories = await Store.ALL_CATEGORIES
  const category = allCategories.find((category) => category.slug === slug)

  if (category) {
    return category
  } else {
    throw 'Category not found.'
  }
}

/* -------------------------------------------
  getAllTags : ALL_TAGS を返却
  getAllTagPaths : ALL_TAGS から slug を抜き出して返却
  getTagPosts : 受け取った slug に一致する記事を ALL_POSTS から返却
  getTag : 受け取った slug に一致するタグを ALL_TAGS から返却
------------------------------------------- */
export const getAllTags = async (): Promise<TagType[]> => {
  const allTags = await Store.ALL_TAGS

  return allTags
}

export const getAllTagPaths = async (): Promise<{ params: { slug: string } }[]> => {
  const allTags = await Store.ALL_TAGS

  return allTags.map((tag) => {
    return {
      params: {
        slug: tag.slug,
      },
    }
  })
}

export const getTagPosts = async (slug: string): Promise<PostType[]> => {
  const allPosts = await Store.ALL_POSTS
  const tagPosts = allPosts.filter((post) => {
    const isFound = post.tags.some((tag) => {
      return tag.slug === slug
    })
    return isFound
  })

  return tagPosts
}

export const getTag = async (slug: string): Promise<TagType> => {
  const allTags = await Store.ALL_TAGS
  const tag = allTags.find((tag) => tag.slug === slug)

  if (tag) {
    return tag
  } else {
    throw 'Tag not found.'
  }
}
