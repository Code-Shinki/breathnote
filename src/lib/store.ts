/* ＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/

  LIB -> STORE

  - microCMS から非同期処理で取得したデータを保存

  - CONFIG : サイト設定 (オブジェクト)
  - ALL_POSTS : 記事一覧 (配列)
  - ALL_CATEGORIES : カテゴリ一覧 (配列)
  - ALL_TAGS : タグ一覧 (配列)

＿/ ＿/ ＿/ ＿/ ＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/＿/ */

import { AllCategoriesType, AllConfigType, AllPostsType, AllTagsType } from 'lib/types'

export const CONFIG = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllConfigType = await fetch('https://breathnote.microcms.io/api/v1/settings/config', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData
})()

export const ALL_POSTS = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllPostsType = await fetch('https://breathnote.microcms.io/api/v1/posts', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData.contents
})()

export const ALL_CATEGORIES = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllCategoriesType = await fetch('https://breathnote.microcms.io/api/v1/categories', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData.contents
})()

export const ALL_TAGS = (async () => {
  const key = {
    headers: { 'X-API-KEY': process.env.API_KEY as string },
  }
  const fetchData: AllTagsType = await fetch('https://breathnote.microcms.io/api/v1/tags', key)
    .then((res) => res.json())
    .catch((err) => console.error(err))

  return fetchData.contents
})()
