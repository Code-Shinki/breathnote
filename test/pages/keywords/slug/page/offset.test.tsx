import { cleanup, render, screen } from '@testing-library/react'
import React from 'react'
import Keyword from '../../../../../src/pages/keywords/[slug]/page/[offset]'
import { getKeywordPosts } from '../../../../../src/scripts/getter'
import { ALL_KEYWORDS } from '../../../../../src/scripts/store'

describe(`KEYWORD ARTICLE PAGE 4/7 (pages/keywords/[slug]/page/[offset].tsx)`, () => {
  afterEach(() => {
    cleanup
  })

  test('snapshot', async () => {
    const offset = 4
    const keyword = (await ALL_KEYWORDS).contents.slice(-1)[0]
    const posts = (await getKeywordPosts(keyword.slug)).slice(-6)
    const allPostLength = 42
    const option = {
      pageType: 'keyword',
      fullPath: `https://blog.shinki.net/keywords/${keyword.slug}/page/${offset}`,
    }
    const { asFragment } = render(
      <Keyword posts={posts} allPostLength={allPostLength} keyword={keyword} option={option} offset={offset} />
    )
    const tree = asFragment()
    expect(tree).toMatchSnapshot()
  })

  test('integration', async () => {
    const offset = 4
    const keyword = (await ALL_KEYWORDS).contents.slice(-1)[0]
    const posts = (await getKeywordPosts(keyword.slug)).slice(-6)
    const allPostLength = 42
    const option = {
      pageType: 'keyword',
      fullPath: `https://blog.shinki.net/keywords/${keyword.slug}/page/${offset}`,
    }
    render(<Keyword posts={posts} allPostLength={allPostLength} keyword={keyword} option={option} offset={offset} />)

    // ヘッダーが存在する
    expect(screen.getByRole('banner')).toBeInTheDocument()
    // メインコンテンツが存在する
    expect(screen.getByRole('main')).toBeInTheDocument()
    // サイドコンテンツ (SHARE, CONTENTS, MORE) が存在する
    expect(screen.getAllByRole('complementary')).toHaveLength(3)
    // フッターが存在する
    expect(screen.getByRole('contentinfo')).toBeInTheDocument()

    // h1属性が1個だけ存在する
    expect(screen.getByRole('heading', { level: 1 })).toBeInTheDocument()
    // ページネーションが存在する
    expect(screen.getByRole('navigation', { name: 'pagination' })).toBeInTheDocument()
  })
})
