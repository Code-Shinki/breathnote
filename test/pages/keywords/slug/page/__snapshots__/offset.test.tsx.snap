// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KEYWORD ARTICLE PAGE 4/7 (pages/keywords/[slug]/page/[offset].tsx) snapshot 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <h1
      class="title"
    >
      Next.js
    </h1>
    <p
      class="counter"
    >
      - 84 Articles -
    </p>
    <p
      class="description"
    >
      Vercel社がOSSとして開発を進めているReactベースのフロントエンドフレームワーク
    </p>
  </div>
  <section
    class="root"
  >
    <article
      class="article"
    >
      <div
        class="thumbnail"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iODUwIiBoZWlnaHQ9IjQ0NSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="JamstackなNext.jsブログにプレビューモードを実装する"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </div>
      <time
        class="date"
        datetime="2021-02-23T14:16:00.797Z"
        itemprop="datepublished"
      >
        2021 / 02 / 23
      </time>
      <h2
        class="title"
      >
        <a
          href="/posts/nextjs-preview-mode"
        >
          JamstackなNext.jsブログにプレビューモードを実装する
        </a>
      </h2>
      <ul
        class="keywords"
      >
        <li>
          Next.js
        </li>
        <li>
          Jamstack
        </li>
      </ul>
    </article>
  </section>
  <nav
    aria-label="pagination"
    class="nav"
  >
    <ul
      class="ul"
    >
      <li
        class="li false"
      >
        <a
          class="link"
          href="/keywords/nextjs"
        >
          1
        </a>
      </li>
      <li
        class="dotted"
      >
        ...
      </li>
      <li
        class="li false"
      >
        <a
          class="link"
          href="/keywords/nextjs/page/3"
        >
          3
        </a>
      </li>
      <li
        class="li li-current"
      >
        <a
          class="link"
          href="/keywords/nextjs/page/4"
        >
          4
        </a>
      </li>
      <li
        class="li false"
      >
        <a
          class="link"
          href="/keywords/nextjs/page/5"
        >
          5
        </a>
      </li>
      <li
        class="dotted"
      >
        ...
      </li>
      <li
        class="li false"
      >
        <a
          class="link"
          href="/keywords/nextjs/page/7"
        >
          7
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
